// @embroider/core@3.3.0/src/html-entrypoint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTMLEntrypoint = void 0;
const shared_internals_1 = require("@embroider/shared-internals");
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const jsdom_1 = require("jsdom");
const partition_1 = __importDefault(require("lodash/partition"));
const zip_1 = __importDefault(require("lodash/zip"));
const html_placeholder_1 = __importDefault(require("./html-placeholder"));
class HTMLEntrypoint {
    constructor(pathToVanillaApp, rootURL, publicAssetURL, filename) {
        this.pathToVanillaApp = pathToVanillaApp;
        this.rootURL = rootURL;
        this.publicAssetURL = publicAssetURL;
        this.filename = filename;
        this.placeholders = new Map();
        this.modules = [];
        this.scripts = [];
        this.styles = [];
        this.dom = new jsdom_1.JSDOM((0, fs_extra_1.readFileSync)((0, path_1.join)(this.pathToVanillaApp, this.filename), 'utf8'));
        for (let tag of this.handledStyles()) {
            let styleTag = tag;
            let href = styleTag.href;
            if (!isAbsoluteURL(href)) {
                let url = this.relativeToApp(href);
                this.styles.push(url);
                let placeholder = new html_placeholder_1.default(styleTag);
                let list = (0, shared_internals_1.getOrCreate)(this.placeholders, url, () => []);
                list.push(placeholder);
            }
        }
        for (let scriptTag of this.handledScripts()) {
            // scriptTag.src include rootURL. Convert it to be relative to the app.
            let src = this.relativeToApp(scriptTag.src);
            if (scriptTag.type === 'module') {
                this.modules.push(src);
            }
            else {
                this.scripts.push(src);
            }
            let placeholder = new html_placeholder_1.default(scriptTag);
            let list = (0, shared_internals_1.getOrCreate)(this.placeholders, src, () => []);
            list.push(placeholder);
        }
    }
    relativeToApp(rootRelativeURL) {
        return rootRelativeURL.replace(this.rootURL, '');
    }
    handledScripts() {
        let scriptTags = [...this.dom.window.document.querySelectorAll('script')];
        let [ignoredScriptTags, handledScriptTags] = (0, partition_1.default)(scriptTags, scriptTag => {
            return !scriptTag.src || scriptTag.hasAttribute('data-embroider-ignore') || isAbsoluteURL(scriptTag.src);
        });
        for (let scriptTag of ignoredScriptTags) {
            scriptTag.removeAttribute('data-embroider-ignore');
        }
        return handledScriptTags;
    }
    handledStyles() {
        let styleTags = [...this.dom.window.document.querySelectorAll('link[rel="stylesheet"]')];
        let [ignoredStyleTags, handledStyleTags] = (0, partition_1.default)(styleTags, styleTag => {
            return !styleTag.href || styleTag.hasAttribute('data-embroider-ignore') || isAbsoluteURL(styleTag.href);
        });
        for (let styleTag of ignoredStyleTags) {
            styleTag.removeAttribute('data-embroider-ignore');
        }
        return handledStyleTags;
    }
    // bundles maps from input asset to a per-variant map of output assets
    render(stats) {
        let insertedLazy = new Set();
        let fastbootVariant = stats.variants.findIndex(v => Boolean(v.runtime === 'fastboot'));
        let supportsFastboot = stats.variants.some(v => v.runtime === 'fastboot' || v.runtime === 'all');
        for (let [src, placeholders] of this.placeholders) {
            let match = stats.entrypoints.get(src);
            if (match) {
                let firstVariant = stats.variants.findIndex((_, index) => Boolean(match.get(index)));
                let matchingBundles = match.get(firstVariant);
                let matchingFastbootBundles = fastbootVariant >= 0 ? match.get(fastbootVariant) || [] : [];
                for (let placeholder of placeholders) {
                    placeholder.clear();
                    if (supportsFastboot && placeholder.isScript()) {
                        // if there is any fastboot involved, we will emit the lazy bundles
                        // right before our first script.
                        let lazyMatch = stats.lazyBundles.get(src);
                        if (lazyMatch && !insertedLazy.has(src)) {
                            insertLazyJavascript(lazyMatch, placeholder, this.rootURL);
                            insertLazyStyles(lazyMatch, placeholder, this.publicAssetURL);
                            insertedLazy.add(src);
                        }
                    }
                    for (let [base, fastboot] of (0, zip_1.default)(matchingBundles, matchingFastbootBundles)) {
                        if (supportsFastboot) {
                            // the fastboot version gets prefixed with the rootURL, which
                            // points to our local build directory, because that's where
                            // fastboot always loads scripts from. If there's no
                            // fastboot-specific variant, fastboot loads the base variant, but
                            // still from rootURL rather than publicAssetURL.
                            fastboot = this.rootURL + (fastboot !== null && fastboot !== void 0 ? fastboot : base);
                        }
                        if (base) {
                            // the browser version gets prefixed with the publicAssetURL
                            // because that's where browsers will load it from
                            base = this.publicAssetURL + base;
                        }
                        if (!base) {
                            // this bundle only exists in the fastboot variant
                            let element = placeholder.start.ownerDocument.createElement('fastboot-script');
                            element.setAttribute('src', fastboot);
                            placeholder.insert(element);
                            placeholder.insertNewline();
                        }
                        else if (!fastboot) {
                            // no fastboot variant
                            placeholder.insertURL(base);
                        }
                        else if (fastboot === base) {
                            // fastboot variant happens to be exactly the same bundle as base
                            // (and publicAssetURL===rootURL), so a plain script tag covers
                            // both
                            placeholder.insertURL(base);
                        }
                        else {
                            // we have both and they differ
                            let element = placeholder.insertURL(base);
                            if (element) {
                                element.setAttribute('data-fastboot-src', fastboot);
                            }
                        }
                    }
                }
            }
            else {
                // no match means keep the original HTML content for this placeholder.
                // (If we really wanted it empty instead, there would be matchingBundles
                // and it would be an empty list.)
                for (let placeholder of placeholders) {
                    placeholder.reset();
                }
            }
        }
        return this.dom.serialize();
    }
}
exports.HTMLEntrypoint = HTMLEntrypoint;
function isAbsoluteURL(url) {
    return /^(?:[a-z]+:)?\/\//i.test(url);
}
// we (somewhat arbitrarily) decide to put the lazy javascript bundles before
// the very first <script> that we have rewritten
function insertLazyJavascript(lazyBundles, placeholder, rootURL) {
    for (let bundle of lazyBundles) {
        if (bundle.endsWith('.js')) {
            let element = placeholder.start.ownerDocument.createElement('fastboot-script');
            // we're using rootURL instead of publicAssetURL here because
            // <fastboot-script> is executed by fastboot, which always loads them from
            // the local build directory. NOT from the publicAssetURL that browsers
            // will use, which could be a CDN.
            element.setAttribute('src', rootURL + bundle);
            placeholder.insert(element);
            placeholder.insertNewline();
        }
    }
}
function insertLazyStyles(lazyBundles, placeholder, publicAssetURL) {
    for (let bundle of lazyBundles) {
        if (bundle.endsWith('.css')) {
            let element = placeholder.start.ownerDocument.createElement('link');
            element.setAttribute('href', publicAssetURL + bundle);
            element.setAttribute('rel', 'stylesheet');
            placeholder.insert(element);
            placeholder.insertNewline();
        }
    }
}
