// @embroider/core/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.virtualContent = exports.ResolverLoader = exports.Resolver = exports.throwOnWarnings = exports.expectWarning = exports.debug = exports.warn = exports.unsupported = exports.todo = exports.jsHandlebarsCompile = exports.WaitForTrees = exports.toBroccoliPlugin = exports.optionsWithDefaults = exports.HTMLEntrypoint = exports.getPackagerCacheDir = exports.getAppMeta = exports.applyVariantToBabelConfig = void 0;
var packager_1 = require("./packager");
Object.defineProperty(exports, "applyVariantToBabelConfig", { enumerable: true, get: function () { return packager_1.applyVariantToBabelConfig; } });
Object.defineProperty(exports, "getAppMeta", { enumerable: true, get: function () { return packager_1.getAppMeta; } });
Object.defineProperty(exports, "getPackagerCacheDir", { enumerable: true, get: function () { return packager_1.getPackagerCacheDir; } });
var html_entrypoint_1 = require("./html-entrypoint");
Object.defineProperty(exports, "HTMLEntrypoint", { enumerable: true, get: function () { return html_entrypoint_1.HTMLEntrypoint; } });
var options_1 = require("./options");
Object.defineProperty(exports, "optionsWithDefaults", { enumerable: true, get: function () { return options_1.optionsWithDefaults; } });
var to_broccoli_plugin_1 = require("./to-broccoli-plugin");
Object.defineProperty(exports, "toBroccoliPlugin", { enumerable: true, get: function () { return __importDefault(to_broccoli_plugin_1).default; } });
var wait_for_trees_1 = require("./wait-for-trees");
Object.defineProperty(exports, "WaitForTrees", { enumerable: true, get: function () { return __importDefault(wait_for_trees_1).default; } });
var js_handlebars_1 = require("./js-handlebars");
Object.defineProperty(exports, "jsHandlebarsCompile", { enumerable: true, get: function () { return js_handlebars_1.compile; } });
var messages_1 = require("./messages");
Object.defineProperty(exports, "todo", { enumerable: true, get: function () { return messages_1.todo; } });
Object.defineProperty(exports, "unsupported", { enumerable: true, get: function () { return messages_1.unsupported; } });
Object.defineProperty(exports, "warn", { enumerable: true, get: function () { return messages_1.warn; } });
Object.defineProperty(exports, "debug", { enumerable: true, get: function () { return messages_1.debug; } });
Object.defineProperty(exports, "expectWarning", { enumerable: true, get: function () { return messages_1.expectWarning; } });
Object.defineProperty(exports, "throwOnWarnings", { enumerable: true, get: function () { return messages_1.throwOnWarnings; } });
var module_resolver_1 = require("./module-resolver");
Object.defineProperty(exports, "Resolver", { enumerable: true, get: function () { return module_resolver_1.Resolver; } });
var resolver_loader_1 = require("./resolver-loader");
Object.defineProperty(exports, "ResolverLoader", { enumerable: true, get: function () { return resolver_loader_1.ResolverLoader; } });
var virtual_content_1 = require("./virtual-content");
Object.defineProperty(exports, "virtualContent", { enumerable: true, get: function () { return virtual_content_1.virtualContent; } });
// this is reexported because we already make users manage a peerDep from some
// other packages (like embroider/webpack and @embroider/compat
__exportStar(require("@embroider/shared-internals"), exports);
