// @embroider/core@3.1.0/src/messages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.debug = exports.unsupported = exports.todo = exports.expectWarning = exports.throwOnWarnings = exports.warn = void 0;
const debug_1 = __importDefault(require("debug"));
const util_1 = require("util");
const todo = (0, debug_1.default)('embroider:todo');
exports.todo = todo;
const unsupported = (0, debug_1.default)('embroider:unsupported');
exports.unsupported = unsupported;
const debug = (0, debug_1.default)('embroider:debug');
exports.debug = debug;
function realWarn(message, params) {
    if (hardFailMode > 0) {
        throw new Error(`Unexpected warning in test suite: ${(0, util_1.format)(message, ...params)}`);
    }
    else {
        console.log('WARNING: ' + (0, util_1.format)(message, ...params));
    }
}
let expectStack = [];
let handled = new WeakSet();
function expectedWarn(message, params) {
    let formattedMessage = (0, util_1.format)(message, ...params);
    for (let pattern of expectStack) {
        if (pattern.test(formattedMessage)) {
            handled.add(pattern);
            return;
        }
    }
    realWarn(message, params);
}
function warn(message, ...params) {
    if (expectStack.length === 0) {
        realWarn(message, params);
    }
    else {
        expectedWarn(message, params);
    }
}
exports.warn = warn;
// for use in our test suites
let hardFailMode = 0;
function throwOnWarnings(hooks) {
    if (hooks) {
        // qunit mode
        hooks.before(() => {
            hardFailMode++;
        });
        hooks.after(() => {
            hardFailMode--;
        });
    }
    else {
        // Jest mode
        beforeAll(() => hardFailMode++);
        afterAll(() => hardFailMode--);
    }
}
exports.throwOnWarnings = throwOnWarnings;
function expectWarning(pattern, fn) {
    expectStack.push(pattern);
    try {
        fn();
    }
    finally {
        expectStack.pop();
    }
    return handled.has(pattern);
}
exports.expectWarning = expectWarning;
